File: app/src/main/java/com/i14yokoro/mldpterminal/TermDisplay.java
package com.i14yokoro.mldpterminal;

import android.util.Log;
import java.util.ArrayList;

public class TermDisplay{
    private static final String TAG = "termDisplay**";
    private static final char LF = '\n';

    private final int displayRowSize;
    private final int displayColumnSize;

    // カーソルの座標
    private int cursorX = 0;
    private int cursorY = 0;

    // 表示中の行数
    private int displaySize = 0;

    // 一番上の行
    private int topRow = 0;
    // 入力された文字を保存する
    private final ArrayList<ArrayList<TextItem>> textList;

    // 文字色(RGB)
    private int charColor = 0x000000;

    // 文字色を変えるエスケープシーケンスを受信したら true
    private boolean colorChange = false;

    // 表示する文字を組み立てる
    private final StringBuilder sb = new StringBuilder();

    // 現在カーソルがある行
    private int currRow = 0;

    // カーソルが画面の外にあれば true
    private boolean isOutOfScreen = false;

    /**
     * @param displayRowSize : 画面の横幅
     * @param displayColumnSize : 画面の縦幅
     */
    public TermDisplay(int displayRowSize, int displayColumnSize){
        this.displayRowSize = displayRowSize;
        this.displayColumnSize = displayColumnSize;

        ArrayList<TextItem> items = new ArrayList<>(displayRowSize);
        textList = new ArrayList<>();
        textList.add(items);
    }

    /**
     * リストに文字を記録する.
     *
     * @param text : 入力文字
     * @param color : 文字色
     */
    public void setTextItem(char text, int color){

        TextItem textItem = new TextItem(text, color);
        int columnSize = getTotalColumns();

        textList.get(columnSize - 1).add(textItem);
        if(text == LF || textList.get(getTotalColumns()-1).size() >= displayRowSize){
            ArrayList<TextItem> items1 = new ArrayList<>(displayRowSize+5);
            textList.add(items1);
        }
    }

    /**
     * y 行目 x 番目の文字をリストから消す．
     *
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     */
    public void deleteTextItem(int x, int y){
        textList.get(y).remove(x);
    }


    /**
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     * @param text : 入力文字
     * @param color : 文字色
     */
    public void changeTextItem(int x, int y, char text, int color){
        if(y < textList.size() && x < textList.get(y).size()) {
            TextItem textItem = new TextItem(text, color);
            textList.get(y).set(x, textItem);
        }
    }

    /**
     * @param y : y 行目
     * @param text : 入力文字
     * @param color : 文字色
     */
    public void addTextItem(int y, char text, int color){
        if(y < textList.size()) {
            TextItem textItem = new TextItem(text, color);
            textList.get(y).add(textItem);
        }
        if((text == LF || textList.get(y).size() >= displayRowSize) && y == getTotalColumns()-1){
            ArrayList<TextItem> items1 = new ArrayList<>(displayRowSize);
            textList.add(items1);
        }
    }

    /**
     *
     */
    public void addEmptyRow(){
        ArrayList<TextItem> items1 = new ArrayList<>();
        textList.add(items1);
    }

    /**
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     * @param text : 入力文字
     * @param color : 文字色
     */
    public void insertTextItem(int x, int y, char text, int color){
        if (y > textList.size() || textList.get(y).size() >= displayRowSize){ //行の長さが最大文字数をこえる
            return;
        }
        if(x <= textList.get(y).size()) { //行の範囲内
            if (getText(displayRowSize-1, y) == LF){ //一番最後がLF
                deleteTextItem(displayRowSize-1, y);
            }
            if(getText(x, y) != LF) { //LFじゃない
                addTextItem(y, text, color);
            } else {
                deleteTextItem(x, y);
                addTextItem(y, text, color);
                addTextItem(y, LF, color);
            }
        }
    }

    /**
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     * @param text : 入力文字
     */
    public void changeText(int x, int y, char text){
        if(y < this.textList.size() && x < this.textList.get(y).size()) {
            this.textList.get(y).get(x).setText(text);
        }
    }

    /**
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     * @return
     */
    public char getText(int x, int y){
        if(y < this.textList.size() && x < this.textList.get(y).size()) {
            return this.textList.get(y).get(x).getText();
        } else {
            return '\u0000';
        }
    }

    /**
     * @param x : 現在の行の x 番目
     * @param y : y 行目
     * @return
     */
    public int getColor(int x, int y){
        if(y < this.textList.size() && x < this.textList.get(y).size()) {
            return this.textList.get(y).get(x).getColor();
        } else {
            return 0x000000;
        }
    }

    /**
     * @param y : y 行目
     * @return
     */
    public int getRowLength(int y){
        if(y >= getTotalColumns()){
            return 0;
        }
        if(textList.size() == 0){
            return 0;
        }
        return textList.get(y).size();
    }

    /**
     * @param y : y 行目
     * @return
     */
    public String getRowText(int y){
        StringBuilder str = new StringBuilder();
        for (int x = 0; x < getRowLength(y); x++){
            str.append(getText(x, y));
        }
        return str.toString();
    }

    public void moveTopRow(int count){
        this.topRow = this.topRow + count;
    }

    /**
     * @return
     */
    public String createDisplay(){
        sb.setLength(0);
        char text;

        int totalColumns = getTotalColumns();
        int displayRowSize = getDisplayRowSize();
        int displayColumnSize = getDisplayColumnSize();

        for(int y = 0; y < displayColumnSize; y++){//これはリストの縦移動用（最大でスクリーンの最大値分移動）

            if((y+topRow >= totalColumns)){ //yがリストよりでかくなったら
                setDisplaySize(y-1);
                return sb.toString();
            }
            for (int x = 0, n = textList.get(y+topRow).size(); x < n; x++){ //xはそのyのサイズまで
                text = textList.get(y+topRow).get(x).getText();
                if (text == LF && y+1 == displayColumnSize){
                    break;
                }
                if (!colorChange){
                    sb.append(text);
                } else {
                    sb.append("<font color=#").append(Integer.toHexString(getColor(x, getTopRow() + y))).append(">").append(text).append("</font>");
                }
                if((x == displayRowSize-1) && text != LF && y+1 != displayColumnSize){
                    sb.append(LF);
                    break;
                }
                if(text == LF){
                    if (y-1 == displayColumnSize){
                        sb.deleteCharAt(sb.length()-1);
                    }
                    break;
                }
            }
            if (y+1 < displayColumnSize && y + getTopRow() < totalColumns - 1 && !getRowText(y + getTopRow()).contains("\n") && textList.get(y + topRow).size() < getDisplayRowSize()) {
                sb.append(LF);
            }
        }
        setDisplaySize(displayColumnSize);

        return sb.toString();
    }

    public int getCursorX() {
        return cursorX;
    }

    public void setCursorX(int cursorX) {
        if(cursorX >= displayRowSize){
            this.cursorX = displayRowSize - 1;
        }else {
            if (cursorX >= textList.get(getCursorY() + topRow).size()) {
                this.cursorX = cursorX % displayRowSize;
                Log.d(TAG, Integer.toString(getCursorX()));
            } else {
                if (cursorX < 0) {
                    this.cursorX = 0;
                } else {
                    this.cursorX = cursorX;
                }
            }
        }
    }

    public int getCursorY() {
        return cursorY;
    }

    public void setCursorY(int cursorY) {
        if(cursorY >= displayColumnSize){
            this.cursorY = displayColumnSize-1;
        } else {
            if (cursorY < 0) {
                this.cursorY = 0;
            } else {
                this.cursorY = cursorY;
            }
        }
        Log.d("changeTop", "top :" + (topRow + this.cursorY));
        setCurrRow(topRow + this.cursorY);
    }

    public int getTotalColumns() {
        return this.textList.size();
    }

    public int getDisplayRowSize() {
        return displayRowSize;
    }

    public int getDisplayColumnSize() {
        return displayColumnSize;
    }

    public int getTopRow() {
        return topRow;
    }

    public void setTopRow(int topRow) {
        this.topRow = topRow;
    }

    public int getDisplaySize() {
        return displaySize;
    }

    public void setDisplaySize(int displaySize) {
        this.displaySize = displaySize;
    }

    public int getCharColor() {
        return charColor;
    }

    public void setCharColor(int defaultColor) {
        this.charColor = defaultColor;
    }

    public void setColorChange(boolean colorChange){
        this.colorChange = colorChange;
    }

    public boolean isColorChange(){
        return this.colorChange;
    }

    public int getCurrRow() {
        return currRow;
    }

    public void setCurrRow(int currRow) {
        this.currRow = currRow;
    }

    public boolean isOutOfScreen() {
        return isOutOfScreen;
    }

    public void setOutOfScreen(boolean outOfScreen) {
        isOutOfScreen = outOfScreen;
    }
}

File: app/src/main/kotlin/com/i14yokoro/mldpterminal/TermDisplay.kt
package com.i14yokoro.mldpterminal

import android.util.Log

import java.util.ArrayList

class TermDisplay(val displayRowSize: Int, val displayColumnSize: Int) {

    var cursorX = 0
        set(cursorX) {
            if (cursorX >= displayRowSize) {
                field = displayRowSize - 1
            } else {
                if (cursorX >= textList[cursorY + topRow].size) {
                    field = cursorX % displayRowSize
                    Log.d(TAG, Integer.toString(cursorX))
                } else {
                    field = if (cursorX < 0) {
                        0
                    } else {
                        cursorX
                    }
                }
            }
        }
    var cursorY = 0
        set(cursorY) {
            field = if (cursorY >= displayColumnSize) {
                displayColumnSize - 1
            } else {
                if (cursorY < 0) {
                    0
                } else {
                    cursorY
                }
            }
            Log.d("changeTop", "top :" + Integer.toString(topRow + this.cursorY))
            currRow = topRow + this.cursorY
        }

    var displaySize = 0

    var topRow = 0
    private val textList: ArrayList<ArrayList<TextItem>>

    var defaultColor = 0x000000

    var isColorChange = false

    private val sb = StringBuilder()

    var currRow = 0

    var isOutOfScreen = false

    val totalColumns: Int
        get() = this.textList.size

    init {

        val items = ArrayList<TextItem>(displayRowSize)
        textList = ArrayList()
        textList.add(items)
    }

    fun setTextItem(text: Char, color: Int) {

        val textItem = TextItem(text, color)
        val columnSize = totalColumns

        textList[columnSize - 1].add(textItem)
        if (text == LF || textList[totalColumns - 1].size >= displayRowSize) {
            val items1 = ArrayList<TextItem>(displayRowSize + 5)
            textList.add(items1)
        }
    }

    fun deleteTextItem(x: Int, y: Int) {
        textList[y].removeAt(x)
    }

    fun changeTextItem(x: Int, y: Int, text: Char, color: Int) {
        if (y < textList.size && x < textList[y].size) {
            val textItem = TextItem(text, color)
            textList[y][x] = textItem
        }
    }

    fun addTextItem(y: Int, text: Char, color: Int) {
        if (y < textList.size) {
            val textItem = TextItem(text, color)
            textList[y].add(textItem)
        }
        if ((text == LF || textList[y].size >= displayRowSize) && y == totalColumns - 1) {
            val items1 = ArrayList<TextItem>(displayRowSize)
            textList.add(items1)
        }
    }

    fun addEmptyRow() {
        val items1 = ArrayList<TextItem>()
        textList.add(items1)
    }

    fun insertTextItem(x: Int, y: Int, text: Char, color: Int) {
        if (y > textList.size || textList[y].size >= displayRowSize) { //行の長さが最大文字数をこえる
            return
        }
        if (x <= textList[y].size) { //行の範囲内
            if (getText(displayRowSize - 1, y) == LF) { //一番最後がLF
                deleteTextItem(displayRowSize - 1, y)
            }
            if (getText(x, y) != LF) { //LFじゃない
                addTextItem(y, text, color)
            } else {
                deleteTextItem(x, y)
                addTextItem(y, text, color)
                addTextItem(y, LF, color)
            }
        }
    }

    fun changeText(x: Int, y: Int, text: Char) {
        if (y < this.textList.size && x < this.textList[y].size) {
            this.textList[y][x].text = text
        }
    }

    private fun getText(x: Int, y: Int): Char {
        return if (y < this.textList.size && x < this.textList[y].size) {
            this.textList[y][x].text
        } else {
            '\u0000'
        }
    }

    private fun getColor(x: Int, y: Int): Int {
        return if (y < this.textList.size && x < this.textList[y].size) {
            this.textList[y][x].color
        } else {
            0x000000
        }
    }

    fun addTopRow(count: Int) {
        this.topRow = this.topRow + count
    }

    fun createDisplay(): String {
        sb.setLength(0)
        var text: Char

        val totalColumns = totalColumns
        val displayRowSize = displayRowSize
        val displayColumnSize = displayColumnSize

        for (y in 0 until displayColumnSize) {//これはリストの縦移動用（最大でスクリーンの最大値分移動）

            if (y + topRow >= totalColumns) { //yがリストよりでかくなったら
                displaySize = y - 1
                return sb.toString()
            }
            var x = 0
            val n = textList[y + topRow].size
            while (x < n) { //xはそのyのサイズまで
                text = textList[y + topRow][x].text
                if (text == LF && y + 1 == displayColumnSize) {
                    break
                }
                if (!isColorChange) {
                    sb.append(text)
                } else {
                    sb.append("<font color=#").append(Integer.toHexString(getColor(x, topRow + y))).append(">").append(text).append("</font>")
                }
                if (x == displayRowSize - 1 && text != LF && y + 1 != displayColumnSize) {
                    sb.append(LF)
                    break
                }
                if (text == LF) {
                    if (y - 1 == displayColumnSize) {
                        sb.deleteCharAt(sb.length - 1)
                    }
                    break
                }
                x++
            }
            if (y + 1 < displayColumnSize && y + topRow < totalColumns - 1 && !getRowText(y + topRow).contains("\n") && textList[y + topRow].size < displayRowSize) {
                sb.append(LF)
            }
        }
        displaySize = displayColumnSize

        return sb.toString()
    }

    fun getRowLength(y: Int): Int {
        if (y >= totalColumns) {
            return 0
        }
        return if (textList.size == 0) {
            0
        } else textList[y].size
    }

    fun getRowText(y: Int): String {
        val str = StringBuilder()
        for (x in 0 until getRowLength(y)) {
            str.append(getText(x, y))
        }
        return str.toString()
    }

    companion object {
        private const val TAG = "termDisplay**"
        private const val LF = '\n'
    }
}
